pkg/internal/controller/controller.go

```diff
@@ -144,6 +167,38 @@ func (c *Controller[request]) NeedLeaderElection() bool {
 	return *c.LeaderElected
 }
 
+// Warmup implements the manager.WarmupRunnable interface.
+func (c *Controller[request]) Warmup(ctx context.Context) error {
+	if c.NeedWarmup == nil || !*c.NeedWarmup {
+		return nil
+	}
+	return c.startEventSources(ctx)
+}
+
+// DidFinishWarmup implements the manager.WarmupRunnable interface.
+func (c *Controller[request]) DidFinishWarmup(ctx context.Context) bool {
```

Comment:
Probably dead code, but should we handle the `if c.NeedWarmup == nil || !*c.NeedWarmup` case here in some way specifically?

Reply:
Hm. Maybe we have to.

Controller will always be added to the Warmup runnable group.

If Warmup is a no-op, DidFinishWarmup will never return, right?

Which means the manager will get stuck when starting the warmup runnable group

Reply:
Yeah we need some way to short-circuit this code if the warmup is not enabled

Reply:
Yes, this was an oversight. I've added a fix with a UT [here](http://github.com/kubernetes-sigs/controller-runtime/pull/3192/commits/667bb0318710617e96f827b5da631c9c774fab23) that will fail without this short circuit.